// @qte/nest-border-patrol@0.3.0/dist/border-patrol-core.module.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "BorderPatrolCoreModule", {
    enumerable: true,
    get: function() {
        return BorderPatrolCoreModule;
    }
});
const _common = require("@nestjs/common");
const _core = require("@nestjs/core");
const _borderpatrolconstants = require("./border-patrol.constants");
const _borderpatrolfilter = require("./border-patrol.filter");
const _borderpatrolexplorerservice = require("./border-patrol-explorer.service");
function _ts_decorate(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for(var i = decorators.length - 1; i >= 0; i--)if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}
let BorderPatrolCoreModule = class BorderPatrolCoreModule {
    static forRootAsync(options) {
        const optionsProvider = {
            provide: _borderpatrolconstants.MODULE_OPTIONS_KEY,
            useFactory: options.useFactory,
            inject: options.inject
        };
        const filterProvider = {
            provide: _core.APP_FILTER,
            inject: [
                _borderpatrolconstants.MODULE_OPTIONS_KEY
            ],
            useFactory: (options)=>{
                if (options.useDefaultFilter || options.useDefaultFilter === undefined) {
                    return new _borderpatrolfilter.BorderPatrolExceptionFilter();
                }
                return null;
            }
        };
        return {
            imports: [
                _core.DiscoveryModule
            ],
            exports: [],
            module: BorderPatrolCoreModule,
            providers: [
                optionsProvider,
                filterProvider,
                _borderpatrolexplorerservice.BorderPatrolExplorerService
            ]
        };
    }
};
BorderPatrolCoreModule = _ts_decorate([
    (0, _common.Global)(),
    (0, _common.Module)({})
], BorderPatrolCoreModule);
