// @qte/nest-border-patrol@0.3.0/dist/border-patrol-explorer.service.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "BorderPatrolExplorerService", {
    enumerable: true,
    get: function() {
        return BorderPatrolExplorerService;
    }
});
const _common = require("@nestjs/common");
const _core = require("@nestjs/core");
const _zodtots = require("zod-to-ts");
const _nodepath = /*#__PURE__*/ _interop_require_default(require("node:path"));
const _borderpatrolconstants = require("./border-patrol.constants");
const _promises = /*#__PURE__*/ _interop_require_default(require("node:fs/promises"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {};
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function _ts_decorate(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for(var i = decorators.length - 1; i >= 0; i--)if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function _ts_metadata(k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
}
let BorderPatrolExplorerService = class BorderPatrolExplorerService {
    discoveryService;
    metadataScanner;
    logger;
    reflector;
    constructor(discoveryService, metadataScanner){
        this.discoveryService = discoveryService;
        this.metadataScanner = metadataScanner;
        this.logger = new _common.Logger(BorderPatrolExplorerService.name);
        this.reflector = new _core.Reflector();
    }
    async onModuleInit() {
        const outEnv = process.env.BORDER_PATROL_GATEWAY_TYPES;
        if (!outEnv) {
            return;
        }
        const wsPackageConstants = await Promise.resolve().then(()=>/*#__PURE__*/ _interop_require_wildcard(require("@nestjs/websockets/constants"))).then((mod)=>{
            return {
                GATEWAY_METADATA: mod.GATEWAY_METADATA,
                MESSAGE_MAPPING_METADATA: mod.MESSAGE_MAPPING_METADATA,
                MESSAGE_METADATA: mod.MESSAGE_METADATA
            };
        }).catch((e)=>null);
        if (!wsPackageConstants) {
            throw new Error("@nestjs/websockets not installed");
        }
        const outPath = _nodepath.default.join(process.cwd(), outEnv);
        if (_nodepath.default.extname(outPath) !== ".ts") {
            throw new Error("Path must be a .ts file");
        }
        const gateways = this.discoveryService.getProviders().filter((wrapper)=>wrapper.metatype && this.reflector.get(wsPackageConstants.GATEWAY_METADATA, wrapper.metatype));
        const gatewayHandlers = gateways.flatMap((gateway)=>{
            const instancePrototype = Object.getPrototypeOf(gateway.instance);
            const methodNames = this.metadataScanner.getAllMethodNames(instancePrototype);
            return methodNames.map((methodName)=>{
                const method = instancePrototype[methodName];
                const isMessageMapping = Reflect.getMetadata(wsPackageConstants.MESSAGE_MAPPING_METADATA, method);
                if (!isMessageMapping) {
                    return null;
                }
                const event = Reflect.getMetadata(wsPackageConstants.MESSAGE_METADATA, method);
                const configuration = Reflect.getMetadata(_borderpatrolconstants.BORDER_CONFIGURATION_KEY, method);
                const dataType = (0, _zodtots.printNode)((0, _zodtots.zodToTs)(configuration.body).node);
                return {
                    event,
                    dataZodSchema: configuration.body,
                    dataTsString: dataType
                };
            });
        }).filter((handler)=>!!handler).sort((a, b)=>a.event.localeCompare(b.event));
        const gatewayMessageTsType = "type ClientMessage = " + gatewayHandlers.map((handler)=>`{
  event: "${handler.event}";
  data: ${handler.dataTsString.replaceAll("\n", "\n\t")};
}
`).join(" | \n");
        _promises.default.mkdir(_nodepath.default.dirname(outPath), {
            recursive: true
        }).then(()=>{
            _promises.default.writeFile(outPath, gatewayMessageTsType, "utf-8");
        });
    }
};
BorderPatrolExplorerService = _ts_decorate([
    (0, _common.Injectable)(),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", [
        typeof _core.DiscoveryService === "undefined" ? Object : _core.DiscoveryService,
        typeof _core.MetadataScanner === "undefined" ? Object : _core.MetadataScanner
    ])
], BorderPatrolExplorerService);
