// @qte/nest-border-patrol@0.3.0/dist/index.js
// Old stuff
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    createBorder: function() {
        return _utility.createBorder;
    },
    createWsBorder: function() {
        return _utility.createWsBorder;
    },
    UseBorder: function() {
        return _borderpatroldecorators.UseBorder;
    },
    extendBorder: function() {
        return _zodopenapi.extendApi;
    },
    BorderPatrolModule: function() {
        return _borderpatrolmodule.BorderPatrolModule;
    },
    BorderPatrolException: function() {
        return _borderpatrolexception.BorderPatrolException;
    },
    HttpBorder: function() {
        return _HttpBorder.HttpBorder;
    },
    UseHttpBorder: function() {
        return _HttpBorder.UseHttpBorder;
    }
});
const _utility = require("./utility/index");
const _borderpatroldecorators = require("./border-patrol.decorators");
const _zodopenapi = require("@anatine/zod-openapi");
const _borderpatrolmodule = require("./border-patrol.module");
const _borderpatrolexception = require("./border-patrol.exception");
const _HttpBorder = require("./HttpBorder/index");
