// @slonik/utilities@46.8.0/dist/utilities/parseDsn.js
"use strict";
// cspell:ignore sslrootcert, sslcert
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseDsn = void 0;
const errors_1 = require("@slonik/errors");
const node_fs_1 = require("node:fs");
const zod_1 = require("zod");
const parseDsn = (dsn) => {
    if (dsn.trim() === '') {
        return {};
    }
    const url = new URL(dsn);
    const connectionOptions = {};
    if (url.host) {
        connectionOptions.host = decodeURIComponent(url.hostname);
    }
    else if (url.searchParams.has('host')) {
        const host = url.searchParams.get('host');
        if (typeof host === 'string' && host) {
            connectionOptions.host = host;
        }
    }
    if (url.port) {
        connectionOptions.port = Number(url.port);
    }
    if (url.pathname && url.pathname !== '/') {
        connectionOptions.databaseName = decodeURIComponent(url.pathname.split('/')[1]);
    }
    if (url.username) {
        connectionOptions.username = decodeURIComponent(url.username);
    }
    if (url.password) {
        connectionOptions.password = decodeURIComponent(url.password);
    }
    const searchParameters = zod_1.z
        .object({
        application_name: zod_1.z.string().optional(),
        options: zod_1.z.string().optional(),
        sslcert: zod_1.z
            .string()
            .optional()
            .describe('Specifies the file name of the client SSL certificate.'),
        sslkey: zod_1.z
            .string()
            .optional()
            .describe('Specifies the location for the secret key used for the client certificate.'),
        sslmode: zod_1.z.enum(['disable', 'no-verify', 'require']).optional(),
        sslrootcert: zod_1.z
            .string()
            .optional()
            .describe('Specifies the name of a file containing SSL certificate authority (CA) certificate(s).'),
    })
        .parse(Object.fromEntries(url.searchParams));
    if (searchParameters.application_name) {
        connectionOptions.applicationName = searchParameters.application_name;
    }
    if (searchParameters.options) {
        connectionOptions.options = searchParameters.options;
    }
    if (searchParameters.sslmode) {
        connectionOptions.sslMode = searchParameters.sslmode;
    }
    let sslCert;
    let sslKey;
    let sslRootCert;
    if (searchParameters.sslcert) {
        try {
            sslCert = (0, node_fs_1.readFileSync)(searchParameters.sslcert, 'utf8');
        }
        catch {
            throw new errors_1.UnexpectedStateError('Failed to read SSL certificate file.');
        }
    }
    if (searchParameters.sslkey) {
        try {
            sslKey = (0, node_fs_1.readFileSync)(searchParameters.sslkey, 'utf8');
        }
        catch {
            throw new errors_1.UnexpectedStateError('Failed to read SSL key file.');
        }
    }
    if (searchParameters.sslrootcert) {
        try {
            sslRootCert = (0, node_fs_1.readFileSync)(searchParameters.sslrootcert, 'utf8');
        }
        catch {
            throw new errors_1.UnexpectedStateError('Failed to read SSL root certificate file.');
        }
    }
    if (sslCert || sslKey || sslRootCert) {
        if ((sslCert && !sslKey) || (!sslCert && sslKey)) {
            throw new errors_1.UnexpectedStateError('Both sslcert and sslkey must be provided together.');
        }
        connectionOptions.ssl = {
            ca: sslRootCert,
            cert: sslCert,
            key: sslKey,
            rejectUnauthorized: searchParameters.sslmode !== 'no-verify',
        };
    }
    return connectionOptions;
};
exports.parseDsn = parseDsn;
