// slonik@46.2.0/dist/connectionMethods/query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.query = void 0;
const executeQuery_1 = require("../routines/executeQuery");
const errors_1 = require("@slonik/errors");
const executionRoutine = async (finalConnection, finalSql, finalValues) => {
    const result = await finalConnection.query(finalSql, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    finalValues);
    const fields = [];
    if (result.fields) {
        for (const field of result.fields) {
            fields.push({
                dataTypeId: field.dataTypeId,
                name: field.name,
            });
        }
    }
    return {
        command: result.command,
        fields,
        notices: result.notices ?? [],
        rowCount: result.rowCount || 0,
        rows: (result.rows || []),
        type: 'QueryResult',
    };
};
const query = async (connectionLogger, connection, clientConfiguration, slonikSql, inheritedQueryId) => {
    try {
        return await (0, executeQuery_1.executeQuery)(connectionLogger, connection, clientConfiguration, slonikSql, inheritedQueryId, executionRoutine);
    }
    catch (error) {
        if (error instanceof errors_1.SlonikError) {
            connection.events.emit('error', error);
        }
        throw error;
    }
};
exports.query = query;
