// @qte/nest-border-patrol@0.3.0/dist/HttpBorder/HttpBorder.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "HttpBorder", {
    enumerable: true,
    get: function() {
        return HttpBorder;
    }
});
const _HttpBorderResponse = require("./HttpBorderResponse");
let HttpBorder = class HttpBorder {
    bodySchema;
    paramsSchema;
    querySchema;
    responsesSchema;
    constructor(options){
        this.bodySchema = options.requestBody || undefined;
        this.paramsSchema = options.parameters?.path || {};
        this.querySchema = options.parameters?.query || {};
        this.responsesSchema = options.responses;
    }
    createResponse(statusCode, body) {
        return new _HttpBorderResponse.HttpBorderResponse({
            statusCode,
            body
        });
    }
    getBodySchema() {
        return this.bodySchema;
    }
    getParamsSchema() {
        return this.paramsSchema;
    }
    getQuerySchema() {
        return this.querySchema;
    }
    getResponseSchema(code) {
        return this.responsesSchema[code];
    }
    getResponsesSchema() {
        return this.responsesSchema;
    }
};
